shimmingtoolbox/prepare_fieldmap.py

```diff
@@ -97,28 +108,54 @@ def prepare_fieldmap(list_nii_phase, echo_times, mag, unwrapper='prelude', mask=
-
-    # Run the unwrapper
-    phasediff_unwrapped = unwrap_phase(nii_phasediff, unwrapper=unwrapper, mag=mag, mask=mask,
-                                       fname_save_mask=fname_save_mask)
+        # If it's 4d (i.e. there are timepoints)
+        if len(phasediff_unwrapped.shape) == 4:
+            phasediff_unwrapped = correct_2pi_offset(phasediff_unwrapped, mag, mask, VALIDITY_THRESHOLD)
 
-    # If it's 4d (i.e. there are timepoints)
-    if len(phasediff_unwrapped.shape) == 4:
-        phasediff_unwrapped = correct_2pi_offset(phasediff_unwrapped, mag, mask, VALIDITY_THRESHOLD)
-
-    # Divide by echo time
-    fieldmap_rad = phasediff_unwrapped / echo_time_diff  # [rad / s]
-    fieldmap_hz = fieldmap_rad / (2 * math.pi)  # [Hz]
+        # Divide by echo time
+        fieldmap_rad = phasediff_unwrapped / echo_time_diff  # [rad / s]
+        fieldmap_hz = fieldmap_rad / (2 * math.pi)  # [Hz]
 
+    else:
+        # Calculates field map based on multi echo phases by running the prelude unwrapper for each phase individually.
+        unwrapped = [unwrap_phase(list_nii_phase[echo_number], unwrapper=unwrapper, mag=mag, mask=mask,
+                                  fname_save_mask=fname_save_mask) for echo_number in range(len(list_nii_phase))]
+        n_echoes = len(list_nii_phase)  # Number of Echoes
+        n_y = unwrapped[0].shape[0]  # Number of voxels along Y direction (rows) in each volume.
+        n_x = unwrapped[0].shape[1]  # Number of voxels along X direction (columns) in each volume.
+        n_z = unwrapped[0].shape[2]  # Number of slices along Z direction in each volume.
+        list_unwrapped_nii = []
+        for echo in range(n_echoes):
+            list_unwrapped_nii.append(unwrapped[echo])
+        # merges all the phases in the 4th dimension
+        merged_unwrapped_phases = np.stack(list_unwrapped_nii, axis=3)
+        # creates a mask based on the given magnitude image
+        new_mask = mask_threshold(mag - mag.min(), VALIDITY_THRESHOLD * (mag.max() - mag.min()))
+        mean = 0
+        for i_time in range(n_echoes):  # corrects 2pi offset between phases
+            new_mean = np.mean(merged_unwrapped_phases[..., i_time][new_mask])
+            n_offsets_float = (mean - new_mean) / (2 * np.pi)
+            n_offsets = round(n_offsets_float)
+            if n_offsets >= 1:
+                i_unwrapped = merged_unwrapped_phases[..., i_time] + n_offsets * np.pi * 2
+                merged_unwrapped_phases[..., i_time] = i_unwrapped
+            mean = new_mean
+        rad_over_time = np.zeros((n_y, n_x, n_z))
+        for slice in range(0, n_z):
+            for voxel_x in range(n_x):
+                for voxel_y in range(n_y):
+                    Y = merged_unwrapped_phases[voxel_y, voxel_x, slice, :]
+                    X = np.asarray(echo_times)
+                    reg = linregress(X, Y)
+                    rad_over_time[voxel_y, voxel_x, slice] = reg.slope
```

Comment:
Triple `for` loops are usually not optimal for speed, you could time it and consider using `sklearn.linear_model.LinearRegression` which can do multiple regressions at once. Here is an example: https://github.com/shimming-toolbox/shimming-toolbox/blob/4275c51a227e06665f39e3a792cb57b94ee40b6e/shimmingtoolbox/shim/realtime_shim.py#L139

I'm not familiar with `linregress`, maybe it can do multiple regressions at once as well.